### PR TITLE
[FLINK-27206][metrics] Deprecate reflection-based reporter instantiation 

### PR DESCRIPTION
**Pull request by zentol**
Based on #17228.

Metric reporters can currently be instantiated in one of 2 ways:
a) the reporter class is loaded via reflection
b) the reporter factory is loaded via reflection/ServiceLoader (aka, plugins)

All reporters provided by Flink use the factory approach, and it is preferable because it supports plugins. The plugin approach also has been available 1.11, and I think it's fair to deprecate the old approach by now.

### CODE REVIEW

**Review by XComp — CHANGES_REQUESTED** [review]
Thanks @zentol . I found some minor things. The change request is based on the cleanup we can do in `PrometheusReporterEndToEndITCase`

File: flink-core/src/main/java/org/apache/flink/configuration/MetricOptions.java [hunk 1]
```diff
@@ -67,8 +67,8 @@ public class MetricOptions {
                                     + " any of the names in the list will be started. Otherwise, all reporters that could be found in"
                                     + " the configuration will be started.");
 
-    @Documentation.SuffixOption(NAMED_REPORTER_CONFIG_PREFIX)
-    @Documentation.Section(value = Documentation.Sections.METRIC_REPORTERS, position = 1)
+    /** @deprecated use {@link MetricOptions#REPORTER_FACTORY_CLASS} instead. */
+    @Deprecated
     public static final ConfigOption<String> REPORTER_CLASS =
```
[the remark this answers is not in the document]
Comment (zentol) [reply flink-core/src/main/java/org/apache/flink/configuration/MetricOptions.java:72]: they are both deprecated now, and are effectively the same thing. I think the current path is fine; someone using `METRICS_REPORTER_CLASS_SUFFIX` is lead to `REPORTER_CLASS`, which then leads to `REPORTER_FACTORY_CLASS`.
Sure, we could remove 1 jump, but I don't think we'd really get anything out of it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/metrics/ReporterSetup.java [hunk 2]
```diff
@@ -360,6 +361,7 @@ private static List<ReporterSetup> setupReporters(
         return reporterSetups;
     }
 
+    @SuppressWarnings("deprecation")
     private static Optional<MetricReporter> loadReporter(
```
[the remark this answers is not in the document]
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/metrics/ReporterSetup.java:365]: they generally don't have to touch reporters when upgrading. I made sure we log a warning whenever we don't use a factory.

File: flink-runtime/src/main/java/org/apache/flink/runtime/metrics/ReporterSetup.java [hunk 3]
```diff
@@ -462,6 +462,15 @@ private static Optional<MetricReporter> loadViaReflection(
                     alternativeFactoryClassName);
             return loadViaFactory(
                     alternativeFactoryClassName, reporterName, reporterConfig, reporterFactories);
+        } else {
+            LOG.warn(
```
[the remark this answers is not in the document]
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/metrics/ReporterSetup.java]: 👍

File: flink-runtime/src/main/java/org/apache/flink/runtime/metrics/ReporterSetup.java [hunk 4]
```diff
@@ -375,6 +377,15 @@ private static Optional<MetricReporter> loadReporter(
         }
 
         if (reporterClassName != null) {
+            LOG.warn(
+                    "The reporter configuration of '{}' configures the reporter class, which is a deprecated approach to configure reporters'."
```
Comment (XComp) [inline flink-runtime/src/main/java/org/apache/flink/runtime/metrics/ReporterSetup.java:381]:
```suggestion
                    "The reporter configuration of '{}' configures the reporter class, which is a deprecated approach to configure reporters."
```

**Review by XComp — APPROVED** [review]
LGMT 👍

**Review by mas-chen — COMMENTED** [review]
Code changes look good. Can we update the examples in the documentation for reporter setup? Prometheus and PrometheusPushGateway examples should advertise factory class configs

**Pull request merged**